[Clang] [C++26] Implement P1306R5 Expansion Statements

Sirraide opened the pull request:
Update: This has now been split into multiple patches, starting with #169680.

This adds support for [expansion statements](https://wg21.link/P1306R5) (aka `template for`) to Clang. The implementation is based in part on @katzdm’s fork, so I’m adding him as a co-author whenever this is merged. I’ve also gone through open core issues for this feature and integrated support for them.

The main thing that currently *isn’t* supported properly is the computation of the size of an iterating expansion statement: this requires synthesising a lambda in Sema and evaluating it, which seems rather complicated, and I’m unsure as to how to go about that, so currently, we cheat by computing `end - begin` instead. Because of this, I also haven’t added the feature test macro yet.

This is a rather large patch that introduces a new `Decl` and `DeclContext` (that being `ExpansionStmtDecl`; we need a `DeclContext` for this because the body of an expansion statement must be treated as a dependent context prior to expansion), and a number of `Stmt`s that correspond to the different kinds of expansion statements, as well as some auxiliary expressions. I considered splitting this up in some fashion to make it easier to review, but given how interconnected this all is, doing so seems non-trivial to me, so I’ve left it as one patch. 

About ~4200 lines of this patch are just tests, and another large portion is just a lot of the range-based `for` loop code that was shuffled around a bit so I could reuse it for iterating expansion statements.

The memory usage of `CXXExpansionStmt` and friends could probably be optimised a bit, but I don’t think there’s an easy way to do that, and I also doubt that expansion statements are common enough to where allocating a bit more memory than strictly necessary for the unexpanded state really matters (if anything, the expansions are going to use a lot more memory anyway).

There is some weirdness with lifetime-extension of temporaries in destructuring expansion statements in codegen: I had to use *two* `LexicalScope`s to make sure the cleanup is emitted in the right place (with just one it would float up to the end of the enclosing scope); I have no idea why that happens, but it might be related to a bug in our lifetime-extension support that I discovered along the way (see #165182). CC @efriedma-quic, @rjmccall maybe you have an idea what might be happening there.

Sirraide committed:
Enumerating expansion statements mostly work

Sirraide committed:
Fix some instantiation issues and add codegen tests

Sirraide committed:
Support pack expansion in enumerating expansion statements

Sirraide committed:
Iterating expansion statements

Sirraide committed:
Dependent expansion statements

Sirraide committed:
Sema for destructuring expansion statements

Sirraide committed:
Implement CWG 3044

Sirraide committed:
Add codegen tests for destructuring expansion statements

Sirraide committed:
Add tests for break/continue

Sirraide committed:
Disallow (non-local) labels in expansion statements

Sirraide committed:
Disallow case/default statements in expansion statements

Sirraide committed:
Add another case/default test

Sirraide committed:
Add comment explaining why we diagnose this here

Sirraide committed:
Proper support for local labels

Sirraide committed:
Make substatement memory representation less horrible

Sirraide committed:
Update out-of-date comment

Sirraide committed:
Check for unexpanded parameter packs

Sirraide committed:
Add -fexpansion-limit

Sirraide committed:
Add release note

Sirraide committed:
Add StmtPrinter/TextNodeDumper/Serialisation tests

Sirraide committed:
Implement CWG 3061

Sirraide committed:
Add a test for CWG 3048

Sirraide committed:
Add a test for lifetime extension (CWG 3034)

Sirraide committed:
Fix lifetime extension in templates

Sirraide committed:
Remove commented-out code

Sirraide committed:
ASTImporter

Sirraide committed:
Minor fixes

Sirraide committed:
clang-format

Sirraide committed:
Update LanguageExtensions.rst

Sirraide committed:
clang-format, again

zyn0217 commented on clang/lib/Sema/SemaExpand.cpp:
```diff
@@ -0,0 +1,584 @@
+#include "clang/Sema/Sema.h"
+
+#include <clang/Sema/EnterExpressionEvaluationContext.h>
+#include <clang/Sema/Template.h>
```
```suggestion

#include "clang/Sema/EnterExpressionEvaluationContext.h"
#include "clang/Sema/Template.h"
```

Sirraide committed:
Add extension warning test

zyn0217 commented on clang/lib/Sema/SemaExpand.cpp:
```diff
@@ -0,0 +1,584 @@
+                        ->getExprs()
+                        .size();
+
+    return Size;
```
```suggestion
    return cast<CXXExpansionInitListSelectExpr>(
                        Expansion->getExpansionVariable()->getInit())
                        ->getRangeExpr()
                        ->getExprs()
                        .size();
```

Sirraide committed:
Apply code review suggestions

Sirraide committed:
Handle returning from an expansion properly in constant evaluation

Sirraide committed:
Merge branch 'main' into expansion-statements

Sirraide committed:
Actually make this test expand to nothing

cor3ntin reviewed (commented):
Thanks for working on this.
I started to review but I had barely any time to scratch the surface

Sirraide commented on clang/include/clang/Basic/DiagnosticParseKinds.td:
```diff
@@ -445,6 +450,8 @@ def err_unspecified_size_with_static : Error<
 def err_expected_parentheses_around_typename : Error<
   "expected parentheses around type name in %0 expression">;
+def err_expansion_stmt_requires_range : Error<
+  "expansion statement must be range-based">;
```
I think it’s easier to do the former because we diagnose this after parsing the `(...)` part

Sirraide commented on clang/include/clang/Basic/StmtNodes.td:
```diff
@@ -177,6 +185,11 @@ def CoyieldExpr : StmtNode<CoroutineSuspendExpr>;
+// C++26 Expansion statement support expressions
+def CXXExpansionInitListExpr : StmtNode<Expr>;
+def CXXExpansionInitListSelectExpr : StmtNode<Expr>;
+def CXXDestructuringExpansionSelectExpr : StmtNode<Expr>;
```
‘impressive’ is definitely one way to describe it; 8 new AST nodes is... not great. I considered merging all the `CXX...ExpansionStmt` nodes into just `CXXExpansionStmt`, but then that one node would have all of the accessors for all the different variants (even though e.g. `getDecompositionDecl()` doesn’t make sense for an enumerating expansion statement), and you’d have to rely on another enum to figure out what kind of expansion statement you’re dealing with, at which point I felt it made more sense to just make it a separate AST mode. I can try refactoring it to where everything is just in `CXXExpansionStmt` if you think that would simplify things (figuring out an AST representation that is somewhat usable was kind of a headache for me with this feature...)

> merging the SelectExpr nodes - to the extent we need these, the index is implicit from the expansion order. And these things are always expanded, when they are created, when do we need an index?

> For the initializer, can we maybe use a template argument, or a super set thereof?

I agree they are pretty awkward. I was thinking of refactoring this to be similar to how we do pack expansion (i.e. have a global index variable somewhere in Sema that’s managed by some RAII object when we perform expansion).

I have two questions about this part though:
1. As I see it, we still need the index as an expression for iterating expansion statements because we need to evaluate `begin + i`. Is there a good way to do that w/o making `i` a DRE to a template parameter?
2. For enumerating and destructuring expansion statements, we really don’t need the index, but I presume the `ExpansionStmtDecl` would still have to have *some* template parameter? Do we support creating or instantiating a template that has no template parameters? Also, even if we do, we don’t know what kind of expansion statement we’re dealing with when we create the `ExpansionStmtDecl`, though we might be able to create the template parameter list and attach it to the declaration once we know that we need it and set it to `nullptr` before that.

Sirraide replied:
> short of just storing the `CXXExpansionInitListExpr*`/`DecompositionDecl*` as a `void*` and just adding a `bool IsExpansionInitList` member to it

I guess we could make it a `PointerIntPair<PointerUnion<CXXExpansionInitListExpr*, DecompositionDecl*>, 1>` or store them as trailing data?

cor3ntin replied:
Yeah, this seems not great. I am not sure I have a great suggestion - Can we add only one node in this PR? I realize it's a lot of work to split things out, but it might help review go a bit faster... there is a lot here

Sirraide committed:
Amend diagnostic wording

Sirraide committed:
Only move up to the parent context during expansion

Sirraide committed:
Add tests for destructuring using tuple_size

Sirraide commented:
> It’s not entirely clear to me which of the conditions I listed above constitute an ill-formed program, and which are grounds for switching to destructuring instead. GCC and the experimental fork also don’t seem to agree with each other on this.

What I believe would have the best QOI is something like
- if either member `begin()` or `end()` exist (irrespective of whether they’re accessible), it is an iterating expansion statement (albeit perhaps an ill-formed one); otherwise
- if ADL for either `begin()` or `end()` finds a single best viable candidate (simply checking if they exist doesn’t suffice here; as stated above, doing ADL on e.g. anything in namespace `std` will find a number of `begin()`/`end()` overloads), it is an iterating expansion statement (albeit perhaps an ill-formed one); otherwise
- it is a destructuring expansion statement.

This in my opinion would have the least risk of potentially confusing what is probably supposed to be an iterating expansion statement for a destructuring one. To me, it stands to reason that if we can find *any* `begin()` or `end()` member, or a *best viable* ADL `begin()` or `end()`, then the user probably intended for this to be an iterating expansion statement. In particular: 

- If either `begin()` or `end()` is deleted, I would assume that user intent is ‘don’t iterate over this or expand it please’, i.e. it should be an ill-formed iterating expansion statement.
- Similarly, if we can only find one of `begin()` or `end()`, it seems more reasonable to me to assume that the user forgot to provide the other one, rather than assuming that it’s supposed to be a destructuring expansion statement, and that the sole `begin()` or `end()` serves some entirely unrelated purpose. From what I recall in the proposal for this feature, ‘if you don’t get the `begin()`/`end()` protocol exactly right, you silently get destructuring instead’ is precisely something that the designers wanted to avoid.

That said, I’m entirely unsure if any of that matches the intent of the standard here.

erichkeane commented on clang/include/clang/AST/StmtCXX.h:
```diff
@@ -524,6 +525,354 @@ class CoreturnStmt : public Stmt {
+  friend class ASTStmtReader;
+
+  ExpansionStmtDecl *ParentDecl;
+  SourceLocation ForLoc;
```
When is this not just `BeginLoc`?

Sirraide replied:
So `getBeginLoc()` for a `CXXExpansionStmt` currently returns the location of the `template` keyword because I figured that made more sense.

Sirraide replied:
> What do we use the `For` loc for then?

Nothing really; I just included it because we typically include keyword locations in the AST; I guess it’s just there in case someone wants to do something w/ it?

cor3ntin replied:
In case we really need it, we could get from it through the source manager - there cannot be anything else here

erichkeane reviewed (changes requested):
I have only gotten through the AST nodes here, but am already convinced this patch needs to be extensively broken down.  This changes an absurd amount of code, and creates a bunch of AST nodes.  IMO, we should start adding support to these 1 by 1, and perhaps separate parse/sema, as this is never going to otherwise get the review quality it needs.

Sirraide commented:
> I have only gotten through the AST nodes here, but am already convinced this patch needs to be extensively broken down. This changes an absurd amount of code, and creates a bunch of AST nodes. IMO, we should start adding support to these 1 by 1, and perhaps separate parse/sema, as this is never going to otherwise get the review quality it needs.

I considered that too, but I’m candidly not sure how much that’d help. If we were to implement just enumerating expansion statements (which are the simplest to implement), then we’d still need:
- A new `DeclContext` (i.e. `ExpansionStmtDecl`), together with all the miscellaneous changes everywhere to make sure things don’t break horribly as soon as you write an expansion statement.
- *At least* 4 new AST nodes (`ExpansionStmtDecl`, `CXXExpansionStmt`, `CXXExpansionInitListExpr`, `CXXExpansionInstantiationStmt`), and that’s assuming we both want and figure out a way to combine some of them, because currently it’s actually 6 (add `CXXExpansionInitListSelectExpr`, and `CXXEnumeratingExpansionStmt`).

We’d be able to remove the code for iterating and destructuring expansion statements, and that might save us maybe a few 100 LOC if we exclude all the code for the new AST nodes, but the latter part is pretty mechanical imo.

If anyone can think of other ways of splitting this up, please let me know, but the main issue is that this feature requires changing code in a bunch of places for anything to have any change of being functional to the point where we can test it.

As for separating parsing and sema, I don’t think there are that many parser changes in this and the actual parsing is pretty straight-forward (at least to me; perhaps I’ve also just been staring at this for too long..); also, because of how the `for` loop parsing code works, it’d end up being rather awkward if we didn’t also introducing the AST nodes as part of that patch.

Sirraide commented on clang/include/clang/AST/DeclTemplate.h:
```diff
@@ -3343,6 +3343,55 @@ class TemplateParamObjectDecl : public ValueDecl,
+class ExpansionStmtDecl : public Decl, public DeclContext {
+  CXXExpansionStmt *Expansion = nullptr;
+  TemplateParameterList *TParams;
+  CXXExpansionInstantiationStmt *Instantiations = nullptr;
```
So the way I’ve set this up is, if the expansion statement has been expanded, we never instantiate it again and instead only instantiate the expansions (because we’d just be spending time instantiating something that will never be used for anything).

We could probably also most this into `CXXExpansionStmt`; iirc I tried that, but it was ultimately easier to just put it here.

erichkeane replied:
Hmm... I see, so this is actually just a SINGLE instantiation when done in a non-dependent context?  The name threw me a bit.  We wouldn't expect the 'dependent' components to this to actually end up anywhere else/be changeable, right? 

So the process is effectively to finish this context, then 'step back' into it and do a tree-transform/instantiation?

Sirraide replied:
> Hmm... I see, so this is actually just a SINGLE instantiation when done in a non-dependent context? The name threw me a bit. We wouldn't expect the 'dependent' components to this to actually end up anywhere else/be changeable, right?
> So the process is effectively to finish this context, then 'step back' into it and do a tree-transform/instantiation?

I’m not sure I explained that too well earlier, and the terminology here is a bit confusing (at least I keep getting confused by it when trying to explain it, because ‘instantiating’ means 2 different thigns here). 

To rephrase this a bit, let’s distinguish between ‘expanding’ an expansion statement (i.e. instantiating its body N times) from ‘instantiating’ the entire thing (because it just happens to be in a template, and we instantiate the containing template); what I meant was, after it’s been ‘expanded’ (which happens as soon as we parse it if the expansion *size* `N` is not dependent), if we *then* instantiate the containing template, we only ‘instantiate’ the expanded statements (and not e.g. the original expansion-initializer).

In the AST, the `ExpansionStmtDecl` is the top-level node for an expansion statement (it itself is inside a `DeclStmt` because it has to be but that’s irrelevant here). An expansion statement essentially has 2 representations (this is similar to the difference between syntactic/semantic form of initializer lists—I *think*; it’s been a while since I looked at the init list code):

1. the `CXXExpansionStmt` (specifically, one of its derived classes) which contains all the parts needed to perform the expansion, and 
2. the `CXXExpansionInstantiationStmt`, which contains the expanded/desugared AST produced by the expansion. 

The latter is only present after expansion (and it is what’s used for constant evaluation and codegen); additionally, once we’ve performed expansion, we never touch the `CXXExpansionStmt` again, even if parts of it are still dependent (which can happen if the expansion statement is inside a template).

For example, if the user writes
```c++
template <typename T>
void f() {
    template for (auto x : {T(1), T(2)}) {
        T a = x;
    }
}
```
the expansion statement is ‘expanded’ immediately (because even though `{T(1), T(2)}` is instantiation-dependent, the size is known to be `2`), i.e. the template effectively becomes
```c++
template <typename T>
void f() {
    { // Note: This outer ‘compound statement’ is actually an ExpansionStmtDecl whose
      // 'ExpansionInstantiationStmt' contains the following 2 compound statements:
        { T a = T(1); }
        { T a = T(2); }
    }
}
```
(of course, we still preserve e.g. the `{T(1), T(2)}` expression in the AST). If we then instantiate e.g. `f<int>`, the instantiation effectively becomes
```c++
void f<int>() {
    { // Note: This outer ‘compound statement’ is the instantiated ExpansionStmtDecl.
        { int a = int(1); }
        { int a = int(2); }
    }
}
```
Notably, the original expansion-initializer, i.e. `{T(1), T(2)}`, is *not* instantiated (since it isn’t needed anymore); that is, we *don’t* end up w/ e.g. `{int(1), int(2)}` in the AST for `f<int>`, but we instead preserve the original `{T(1), T(2)}`. That is, the `ExpansionStmtDecl` in `f<int>` has both the *original* `CXXExpansionStmt` as written in the template `f`, as well as an ‘instantiation’ of the `ExpansionInstantiationStmt` that was ‘expanded’ when we parsed the template.

Sirraide replied:
> I wonder if we could start on JUST the ExpansionStmtDecl upstreaming. It wouldn't do much besides parsing, but at that point we could add expansions 1 by 1 as we figure out which ones we can need/use.

The main complication w/ that is that because it adds a new `DeclContext`, there are a bunch of DeclContext-related changes in like a dozen files that I had to make, and my worry is that we might introduce either unneeded changes or forget about some of them if we upstream just `ExpansionStmtDecl` w/o a way to test how it’s used. I can extracting all of that and see what we’d end up w/ if you think it’s worth it though.

> I might also suggest trying to un-overload the 'instantiation' phrase here, nad see if we could call it 'expanded' or something like that.

Yeah, it’s definitely annoying that the standard refers to this process of ‘expanding’ as ‘instantiating’. `CXXExpandedExpansionStmt` sounds a bit silly but something like would be an option; let me know if you can think of a better name... 

Ideally, we’d not make this a separate AST node and instead just store all of the expansions in either `ExpansionStmtDecl` or `CXXExpansionStmt`, but both of those are created before we get to the expansion, so at that point we have no idea how many expansions we’ll have, so we can’t exactly allocate trailing memory for it... (just reusing e.g. `CompoundStmt` for this doesn’t work because we need special handling for `break`/`continue` when we codegen/evaluate it).

erichkeane replied:
> > I wonder if we could start on JUST the ExpansionStmtDecl upstreaming. It wouldn't do much besides parsing, but at that point we could add expansions 1 by 1 as we figure out which ones we can need/use.
> 
> The main complication w/ that is that because it adds a new `DeclContext`, there are a bunch of DeclContext-related changes in like a dozen files that I had to make, and my worry is that we might introduce either unneeded changes or forget about some of them if we upstream just `ExpansionStmtDecl` w/o a way to test how it’s used. I can try extracting all of that and see what we’d end up w/ if you think it’s worth it though.
> 

patches that are MOSTLY mechanical for stuff like that is way easier to review.  

> > I might also suggest trying to un-overload the 'instantiation' phrase here, nad see if we could call it 'expanded' or something like that.
> 
> Yeah, it’s definitely annoying that the standard refers to this process of ‘expanding’ as ‘instantiating’. `CXXExpandedExpansionStmt` sounds a bit silly but something like would be an option; let me know if you can think of a better name...

yeah, that is unfortunate but understandable based on their concept.  Perhaps leave it `instantiate/etc` but be REALLY careful with comments here.

> 
> Ideally, we’d not make this a separate AST node and instead just store all of the expansions in either `ExpansionStmtDecl` or `CXXExpansionStmt`, but both of those are created before we get to the expansion, so at that point we have no idea how many expansions we’ll have, so we can’t exactly allocate trailing memory for it... (just reusing e.g. `CompoundStmt` for this doesn’t work because we need special handling for `break`/`continue` when we codegen/evaluate it).


Yeah, that is unfortunate.  Its too bad we couldn't have talked about this before it got standardized and suggested nixing those two:)

Sirraide replied:
Ok, I’ve added some more documentation and examples to the various AST nodes; let me know if this makes the purpose of individual nodes clearer.

cor3ntin replied:
Did you research the possibility of making non-decl decl contexts? 
This is not really the first time we need something like that and it would be nice to be able to say

CXXExpansionStmt : public DeclContext, public Stmt

(alternatively, decoupling template depth tracking from decl context, this is very similar to our lambdas issues @mizvekov )

FYI, I fully realize I'm asking for a pony here, I just want to know if any exploration was done.

In the meantime, maybe we could find a better naming scheme
like
```
  - CXXExpansionStmtDeclContext
     - CXXExpansionStmtPattern
     - CXXExpansionStmtInstantiation
```

Sirraide replied:
> Did you research the possibility of making non-decl decl contexts? This is not really the first time we need something like that and it would be nice to be able to say
> 
> CXXExpansionStmt : public DeclContext, public Stmt


I briefly looked into that, but there are enough places where we do `cast<Decl>(some_decl_context)` or rely on `some_decl_context->getKind()` to return a `DeclKind` to the point that I gave up on that; I have a feeling that that’d be a major refactor.
 
> (alternatively, decoupling template depth tracking from decl context)

I’m candidly not familiar enough with our template engine to know what exactly that would entail.
 
> In the meantime, maybe we could find a better naming scheme like
> 
> ```
>   - CXXExpansionStmtDeclContext
>      - CXXExpansionStmtPattern
>      - CXXExpansionStmtInstantiation
> ```

I’m pretty bad at naming so that’s definitely better than what I came up with. I’ll change it to that instead.

Sirraide commented on clang/include/clang/AST/StmtCXX.h:
```diff
@@ -524,6 +525,354 @@ class CoreturnStmt : public Stmt {
+  }
+};
+
+/// Represents an unexpanded enumerating expansion statement.
```
Sure, I’ll add some examples.

Sirraide commented:
> IF all 4 nodes are required to do ANY work here, I suspect we probably should be combining things. Honestly, this patch is so much work, I don't see myself being able to review this meaningfully in any reasonable amount of time.

Oh yeah, that’s perfectly understandable; I wouldn’t expect a single person to be able to review of it—at least definitely not w/o taking several breaks (which is also why I added a lot of reviewers ;þ).

> I don't have a great suggestion on how to split it up, but I cannot see myself getting through a review on this sufficient to have this accepted/merged as-is.

After thinking about this a bit more, I guess we could make introducing all of the AST nodes a separate patch? But 1. it feels weird to just add a bunch of ‘unused’ AST nodes, and 2. it might get annoying if we then find out that we actually want to change some of them as part of the remaining review process.

Another thing that we could do is: *most* of the changes in `SemaStmt.cpp` are just an NFC refactor of the range-based `for` loop code that extracts a few parts so the iterating expansion statement code can reuse them; I *might* be able to make that a separate patch, though I’m not sure if that would actually be all that helpful.

erichkeane commented:
> > IF all 4 nodes are required to do ANY work here, I suspect we probably should be combining things. Honestly, this patch is so much work, I don't see myself being able to review this meaningfully in any reasonable amount of time.
> 
> Oh yeah, that’s perfectly understandable; I wouldn’t expect a single person to be able to review of it—at least definitely not w/o taking several breaks (which is also why I added a lot of reviewers ;þ).
> 
> > I don't have a great suggestion on how to split it up, but I cannot see myself getting through a review on this sufficient to have this accepted/merged as-is.
> 
> After thinking about this a bit more, I guess we could make introducing all of the AST nodes a separate patch? But 1. it feels weird to just add a bunch of ‘unused’ AST nodes, and 2. it might get annoying if we then find out that we actually want to change some of them as part of the remaining review process.
> 
> Another thing that we could do is: _most_ of the changes in `SemaStmt.cpp` are just an NFC refactor of the range-based `for` loop code that extracts a few parts so the iterating expansion statement code can reuse them; I _might_ be able to make that a separate patch, though I’m not sure if that would actually be all that helpful.

I think a refactor there would be perhaps somewhat helpful, but I would love to see as much of this moved to 'simpler' patches if at all possible, even if they don't implement a whole feature.

Sirraide committed:
Perform overload resolution instead of just ADL

Sirraide committed:
Add a bunch of documentation

Sirraide committed:
Add 'Example:'

Sirraide committed:
Use a C array instead of std::array

Sirraide committed:
Use 'std::tuple' in examples instead of a C array

This should be clearer since C arrays *are* ranges; it's just that
the standard specifically excludes array types from being
expansion-iterable; it should be a bit more obvious that this is
supposed to be destructuring if we use 'std::tuple' as an example.

cor3ntin commented on clang/docs/LanguageExtensions.rst:
```diff
@@ -1784,6 +1784,7 @@ Pack Indexing                                 __cpp_pack_indexing              C
 ``= delete ("should have a reason");``        __cpp_deleted_function           C++26         C++03
 Variadic Friends                              __cpp_variadic_friend            C++26         C++03
 Trivial Relocatability                        __cpp_trivial_relocatability     C++26         C++03
+Expansion Statements                                                           C++26         C++03
```
Can we wait before backporting?

Sirraide replied:
Sure; I can make it C++26 only for now

shafik commented on clang/include/clang/AST/StmtCXX.h:
```diff
@@ -524,6 +525,479 @@ class CoreturnStmt : public Stmt {
+  };
+
+  // Managing the memory for this properly would be rather complicated, and
+  // expansion statements are fairly uncommon, so just allocate space for the
```
Statement like this "expansion statements are fairly uncommon" tend to not hold up well over time.

Sirraide commented on clang/test/SemaCXX/cxx2c-expansion-statements.cpp:
```diff
@@ -0,0 +1,1176 @@
+  template for (auto x : TupleSizeAndGet()) {} // expected-error {{implicit instantiation of undefined template 'std::tuple_element<0, std::TupleSizeAndGet>'}} \
+                                                  expected-note {{in implicit initialization of binding declaration}}
+}
+} // namespace std
```
That seems to just work w/ this patch, but I’ll add it as a test

Sirraide committed:
Merge branch 'main' into expansion-statements

Sirraide committed:
Add NonSFINAEContext

Sirraide committed:
Add test

Sirraide committed:
Diagnose invalid for-range declarations properly

Sirraide committed:
No backporting for now

Sirraide committed:
Remove ForLoc

Sirraide committed:
Rename a bunch of AST nodes

Sirraide commented:
Ok, I’m going to try and split this into a stack of prs now

Sirraide commented:
> Ok, I’m going to try and split this into a stack of prs now

Alright; that’s done now. I’m closing this pr in favour of the patch series starting w/ #169680.
